### ISSUE TITLE
Use additional build information from Simulink

### ISSUE THREAD

**Issue by Marvin-TMG** [CATIA-Systems/FMIKit-Simulink#172]
* Include directories
* Source files
* Librariers
* Defines

**Comment by t-sommer** [comment]
Here's my first take on it. Can you give a90a5d8ccf9984cac63d4215a2d0b99db1796cbd a try, @Marvin-TMG?

**Comment by Marvin-TMG** [comment]
Thanks, I’ll test it on Monday when I’m back in the office.

**Comment by Marvin-TMG** [comment]
I tried it, but had to make some modifications to make it work. The problem is that I use it in simulation mode, not for code generation, thus the build info is defined differently. I also had to remove newline characters and a few minor changes. 

This below is not a generic solution, but it is what worked for me. I am not sure how to do this better, considering that the additional paths depend on how to build the S-Function (i.e. for code generation or simply for simulation).

% custom sources
    custom_include = get_param(gcs, 'SimUserIncludeDirs');
    custom_include = split_paths(custom_include);
    for i = 1:numel(custom_include)
        mex_args{end+1} = ['-I"' strrep(custom_include{i}, newline, '') '"'];
    end
    for i = 1:numel(custom_include)
        mex_args{end+1} = ['-L"' strrep(custom_include{i}, newline, '') '"'];
    end
    
    % custom libraries
    custom_library = get_param(gcs, 'SimUserLibraries');
    custom_library = split_paths(custom_library);
    for i = 1:numel(custom_library)
        mex_args{end+1} = ['-l"' strrep(custom_library{i}, newline, '') '"'];
    end

**Comment by t-sommer** [comment]
The path list in `SimUserIncludeDirs` must be space-separated (not newline-separated, see https://www.mathworks.com/help/simulink/gui/include-directories.html).

**Comment by Marvin-TMG** [comment]
I put this information here for reference: for my specific case I had to modify the lines below to modify the statically linked libraries instead:
`%     for i = 1:numel(libraries)
%         mex_args{end+1} = ['"' libraries{i} '"'];
%     end
    for i = 1:numel(libraries)
        [path, file] = fileparts(libraries{i});
        mex_args{end+1} = ['-L"' path '"'];
    end
    for i = 1:numel(libraries)
        [path, file] = fileparts(libraries{i});
        mex_args{end+1} = ['-l"' file '"'];
    end`

**Comment by t-sommer** [comment]
@Marvin-TMG, when using the `-l` switch, are you passing the filename with or without the extension (`.lib`)?

**Comment by Marvin-TMG** [comment]
Hi Torsten,

The command [path, file, ext] = fileparts(fullfile) is used to remove the extension in my code, so eventhough I specify the full path in the simulink dialog, the name provided to –l is without extension (this is important, otherwise it won’t work).

Kind regards,

Marvin

**Comment by t-sommer** [comment]
@Marvin-TMG, can you give 1d46c10a1616a02ad04721aed711ee30379db86d a try?

**Comment by Marvin-TMG** [comment]
Thanks for the update, I gave it a quick test offline and it seems to be working as expected. Not needing to define Shlwapi.dll is a major improvement.

I have two minor issues still, that I suggest fixing:

·         Separation of build includes/paths is strictly space delimited, as discussed before using line break will work in Matlab, but not with FMIKit. I know this goes against the official Matlab guidelines, but it will be prone to user error, I expect a large portion of the users are not aware of the official delimiter anyway. My suggest is to simply replace line breaks with white space, this won’t do any harm, but will make it more user friendly.

·         Double paths are still being generated, this is quite annoying, especially when using multiple FMUs. For example, the path to FMIKit\include is added once for each new FMU. Parsing the list of paths once, then doing a unique() and writing it back is a safe way to clean this up.

I will get back to you after I tried compiling for the real time platform.

Thank you for your effort!

Kind regards,

Marvin
